src/html_files/analytics.ts

```diff
@@ -0,0 +1,71 @@
+class Rule {
+    name: string;
+    func: Function;
+    good: string;
+    bad: string;
+}
+
+class Rules {
+    data_type: string;
+    pretty_name: string;
+    rules: Array<Rule>;
+}
+
+class RuleOpts {
+    data_type: string;
+    runs: Array<string>;
+    key: string;
+    all_data: any;
+    per_run_data: Map<string, any>;
+}
+
+enum Status {
+    Good = '✅',
+    NotGood = '❌',
+}
+
+class Analytics {
+    name: string;
+    analysis: Array<Finding>;
+}
+
+class Finding {
+    text: string;
+    status: string;
+    recommendation: string;
+
+    constructor(text: string = '', status: Status = Status.NotGood, recommendation: string = '') {
+        if (status == Status.Good) {
+            this.status = '✅';
+        } else {
+            this.status = '❌';
+        }
+        this.text = text;
+        this.recommendation = recommendation;
+    }
+
+    is_good() {
+        this.status == '✅';
+    }
+
+    is_not_good() {
+        this.status == '❌';
+    }
+}
+
+function is_unique_map(values_map) {
+    return new Set([...values_map.values()]).size == 1;
+}
+
+function is_unique_array(values_array) {
+    return new Set(values_array).size == 1;
+}
```

Comment:
These are odd definitions of 'unique'. It suggests that somewhere else you're using the wrong data type if you have to call into something to de-duplicate and check the size.

Reply:
Yeah. We're using arrays everywhere now. We were using arrays and maps. This is cleaner.